Summation introduces spurious symbols

From SO:
https://stackoverflow.com/questions/65094743/how-to-know-the-value-of-automatic-substitutions-by-sympy-after-derivation-of-ex

Code:
```python
from sympy import *
i,b,e,N,B,E = symbols("i b e N B E", cls=Idx)
z =IndexedBase('z')
x =IndexedBase('x')
y =IndexedBase('y')
s2 = Sum((log(Sum(exp(z[b+1,e]*(1/(1+exp(-y[b+1,e])))),[e,1,E])))-(log(Sum(exp(z[b,e]*(1/(1+exp(-y[b,e])))),[e,1,E]))),[b,1,B-1])
pprint(diff(s2,y[1,1]).doit())
```
Output:
```python
In [2]: s2
Out[2]: 
 B - 1                                                              
________                                                            
╲                                                                   
 ╲       ⎛   ⎛  E                     ⎞      ⎛  E                 ⎞⎞
  ╲      ⎜   ⎜_____                   ⎟      ⎜_____               ⎟⎟
   ╲     ⎜   ⎜╲                       ⎟      ⎜╲                   ⎟⎟
    ╲    ⎜   ⎜ ╲                      ⎟      ⎜ ╲                  ⎟⎟
     ╲   ⎜   ⎜  ╲       z[b + 1, e]   ⎟      ⎜  ╲       z[b, e]   ⎟⎟
      ╲  ⎜   ⎜   ╲   ─────────────────⎟      ⎜   ╲   ─────────────⎟⎟
      ╱  ⎜log⎜   ╱        -y[b + 1, e]⎟ - log⎜   ╱        -y[b, e]⎟⎟
     ╱   ⎜   ⎜  ╱    1 + ℯ            ⎟      ⎜  ╱    1 + ℯ        ⎟⎟
    ╱    ⎜   ⎜ ╱    ℯ                 ⎟      ⎜ ╱    ℯ             ⎟⎟
   ╱     ⎜   ⎜╱                       ⎟      ⎜╱                   ⎟⎟
  ╱      ⎜   ⎜‾‾‾‾‾                   ⎟      ⎜‾‾‾‾‾               ⎟⎟
 ╱       ⎝   ⎝e = 1                   ⎠      ⎝e = 1               ⎠⎠
╱                                                                   
‾‾‾‾‾‾‾‾                                                            
 b = 1                                                              

In [3]: s2.diff(y[1, 1]).doit()
Out[3]: 
  ⎛⎧                          -c₃                                       ⎞ ⎛⎧     z[1, 1]                       ⎞   ⎛⎧           
  ⎜⎪                         ℯ                                    c₂    ⎟ ⎜⎪  ─────────────                    ⎟   ⎜⎪           
  ⎜⎪                         ────                            for ℯ   = 1⎟ ⎜⎪       -y[1, 1]                    ⎟   ⎜⎪           
  ⎜⎪                          E                                         ⎟ ⎜⎪  1 + ℯ                            ⎟   ⎜⎪           
  ⎜⎪                                                                    ⎟ ⎜⎪ ℯ             ⋅z[1, 1]            ⎟   ⎜⎪           
- ⎜⎨              c₂                                                    ⎟⋅⎜⎨────────────────────────  for E ≥ 1⎟ + ⎜⎨           
  ⎜⎪             ℯ                            1                         ⎟ ⎜⎪ y[1, 1]        -y[1, 1]           ⎟   ⎜⎪           
  ⎜⎪- ───────────────────────── + ─────────────────────────   otherwise ⎟ ⎜⎪ℯ        + 2 + ℯ                   ⎟   ⎜⎪- ─────────
  ⎜⎪     c₂  c₃  c₂⋅E    c₂  c₃      c₂  c₃  c₂⋅E    c₂  c₃             ⎟ ⎜⎪                                   ⎟   ⎜⎪     c₂  c₃
  ⎜⎪  - ℯ  ⋅ℯ  ⋅ℯ     + ℯ  ⋅ℯ     - ℯ  ⋅ℯ  ⋅ℯ     + ℯ  ⋅ℯ               ⎟ ⎜⎪           0              otherwise⎟   ⎜⎪  - ℯ  ⋅ℯ  
  ⎝⎩                                                                    ⎠ ⎝⎩                                   ⎠   ⎝⎩           

               -c₃                                       ⎞ ⎛⎧    z[B, 1]                           ⎞
              ℯ                                    c₂    ⎟ ⎜⎪ ─────────────                        ⎟
              ────                            for ℯ   = 1⎟ ⎜⎪      -y[B, 1]                        ⎟
               E                                         ⎟ ⎜⎪ 1 + ℯ                                ⎟
                                                         ⎟ ⎜⎪ℯ             ⋅δ   ⋅z[B, 1]           ⎟
   c₂                                                    ⎟⋅⎜⎨                1,B                   ⎟
  ℯ                            1                         ⎟ ⎜⎪───────────────────────────  for E ≥ 1⎟
──────────────── + ─────────────────────────   otherwise ⎟ ⎜⎪   y[B, 1]        -y[B, 1]            ⎟
  c₂⋅E    c₂  c₃      c₂  c₃  c₂⋅E    c₂  c₃             ⎟ ⎜⎪  ℯ        + 2 + ℯ                    ⎟
⋅ℯ     + ℯ  ⋅ℯ     - ℯ  ⋅ℯ  ⋅ℯ     + ℯ  ⋅ℯ               ⎟ ⎜⎪                                      ⎟
                                                         ⎠ ⎝⎩             0               otherwise⎠
```
The symbols C2 and C3 should not be there.

@Maelstrom6 said:

I think you found a bug. SymPy thinks it solved the sum but in reality it hasn't. I don't think this possible to get explicitly since we have no knowledge of the elements of y or z (Try to solve the third inner sum on the first print statement and you'll see what I mean). This is a bug and for anyone that wants to fix this, the issue occurs here. When `e_exp` is None, e does not contain c2 or c3 even though the later code thinks that it does.

The link is to here:
https://github.com/sympy/sympy/blob/master/sympy/concrete/summations.py#L1181-L1187

A possible fix is:
```diff
diff --git a/sympy/concrete/summations.py b/sympy/concrete/summations.py
index a8e083bc6c..c7555cfdc6 100644
--- a/sympy/concrete/summations.py
+++ b/sympy/concrete/summations.py
@@ -1184,14 +1184,13 @@ def eval_sum_symbolic(f, limits):
             if e_exp is not None:
                 e.update(e_exp)
 
-        if e is not None:
-            p = (c1**c3).subs(e)
-            q = (c1**c2).subs(e)
+                p = (c1**c3).subs(e)
+                q = (c1**c2).subs(e)
 
-            r = p*(q**a - q**(b + 1))/(1 - q)
-            l = p*(b - a + 1)
+                r = p*(q**a - q**(b + 1))/(1 - q)
+                l = p*(b - a + 1)
 
-            return Piecewise((l, Eq(q, S.One)), (r, True))
+                return Piecewise((l, Eq(q, S.One)), (r, True))
 
         r = gosper_sum(f, (i, a, b))
```
That looks like a good change but I haven't checked if it gives the correct result. The result it gives is:
```console
$ python wer.py 
  ⎧     z[1, 1]                          ⎧    z[B, 1]                           
  ⎪  ─────────────                       ⎪ ─────────────                        
  ⎪       -y[1, 1]                       ⎪      -y[B, 1]                        
  ⎪  1 + ℯ                               ⎪ 1 + ℯ                                
  ⎪ ℯ             ⋅z[1, 1]               ⎪ℯ             ⋅δ   ⋅z[B, 1]           
  ⎨────────────────────────  for E ≥ 1   ⎨                1,B                   
  ⎪ y[1, 1]        -y[1, 1]              ⎪───────────────────────────  for E ≥ 1
  ⎪ℯ        + 2 + ℯ                      ⎪   y[B, 1]        -y[B, 1]            
  ⎪                                      ⎪  ℯ        + 2 + ℯ                    
  ⎪           0              otherwise   ⎪                                      
  ⎩                                      ⎩             0               otherwise
- ──────────────────────────────────── + ───────────────────────────────────────
            E                                        E                          
          _____                                    _____                        
          ╲                                        ╲                            
           ╲                                        ╲                           
            ╲       z[1, e]                          ╲       z[B, e]            
             ╲   ─────────────                        ╲   ─────────────         
             ╱        -y[1, e]                        ╱        -y[B, e]         
            ╱    1 + ℯ                               ╱    1 + ℯ                 
           ╱    ℯ                                   ╱    ℯ                      
          ╱                                        ╱                            
          ‾‾‾‾‾                                    ‾‾‾‾‾                        
          e = 1                                    e = 1   
```

--- a/sympy/concrete/summations.py
+++ b/sympy/concrete/summations.py
@@ -1161,14 +1161,11 @@ def eval_sum_symbolic(f, limits):
             if e_exp is not None:
                 e.update(e_exp)
 
-        if e is not None:
-            p = (c1**c3).subs(e)
-            q = (c1**c2).subs(e)
-
-            r = p*(q**a - q**(b + 1))/(1 - q)
-            l = p*(b - a + 1)
-
-            return Piecewise((l, Eq(q, S.One)), (r, True))
+                p = (c1**c3).subs(e)
+                q = (c1**c2).subs(e)
+                r = p*(q**a - q**(b + 1))/(1 - q)
+                l = p*(b - a + 1)
+                return Piecewise((l, Eq(q, S.One)), (r, True))
 
         r = gosper_sum(f, (i, a, b))
 

--- a/sympy/concrete/tests/test_sums_products.py
+++ b/sympy/concrete/tests/test_sums_products.py
@@ -1354,6 +1354,10 @@ def test_issue_19379():
     assert Sum(factorial(n)/factorial(n + 2), (n, 1, oo)).is_convergent() is S.true
 
 
+def test_issue_20777():
+    assert Sum(exp(x*sin(n/m)), (n, 1, m)).doit() == Sum(exp(x*sin(n/m)), (n, 1, m))
+
+
 def test__dummy_with_inherited_properties_concrete():
     x = Symbol('x')
 

